### ISSUE TITLE
posAtCoords includes vertical padding only when selection is at the start of inline elements

### ISSUE THREAD

**Issue by hedgepigdaniel** [ProseMirror/prosemirror#1104]
### Issue details

The vertical coordinates of the DOMRect returned by posAtCoords inconsistently do and do not include padding applied to inline elements.

### Steps to reproduce

 - wrap text in inline elements, and apply padding to those elements (e.g. `padding-top: 5px; padding-bottom: 5px;` - useful for changing the height of the background), and DOM like `text<span>textinnode</span>`
 - call `posAtCoords` with a position that is in the middle of a text node - notice that the DOMRect does not include the padding
 - call `posAtCoords` with a position that is at the starting boundary of an inline element (e.g. at the start of an inline decoration). notice that the DOMRect does include padding

### ProseMirror version

1.15.7

### Affected platforms

- [x] Chrome
- [x] Firefox
- [?] Internet Explorer
- [?] Other

**Comment by hedgepigdaniel** [comment]
Still reproducable after the [recent change to `posAtCoords`](https://github.com/ProseMirror/prosemirror-view/commit/250d4337216b5c24ccd4a847cc8e29c13694a33b).

I think there are some incorrect assumptions in the code. [`Range.getClientRects`](https://drafts.csswg.org/cssom-view/#dom-range-getclientrects) does not behave the same as [`Element.getClientRects`](https://drafts.csswg.org/cssom-view/#dom-element-getclientrects). Particularly, the `Range` version includes extra rects which represent the selected parts of text nodes (which do not include padding, unlike the ones for the elements in it).

The problem happens when `singleRect` is called with an element (non text node), and a bias of `-1`. In that case it selects the first rect, which may correspond to a child element rather than text.

Here's some code (which I think works, at least it does in my environment) that takes an element node and a side, and returns a rect for the text at the start or end of the element:

```javascript
function boundaryRect(element, side) {
  let range = document.createRange()

  // Make a range that exactly encompasses the element
  range.setStart(element, 0)
  range.setEndAfter(element)

  // Get the rects in the range
  let rects = range.getClientRects()

  // Perhaps if the element is invisible?
  if (rects.length === 0) return range.getBoundingClientRect()

  // The end is easy - it's always the last rect
  if (side > 0) return rects[rects.length - 1]

  // Not sure if there is always at least two, or if this occurs only when there is padding or some such thing
  if (rects.length === 1) return rects[0]

  // Find the first text node inside the element
  let walker = document.createTreeWalker(range.commonAncestorContainer, 4)
  let firstTextNode = walker.firstChild()

  // If there isn't one, not sure we can do any better
  if (!firstTextNode) return rects[0]

  // Make a range surrounding the text node
  range.setStart(firstTextNode, 0)
  range.setEndAfter(firstTextNode)
  let firstTextNodeRects = range.getClientRects()

  // The first rect should be for the beginning of the text
  return firstTextNodeRects.length > 0 ? firstTextNodeRects[0] : rects[0]
}
```

**Comment by marijnh** [comment]
> I think there are some incorrect assumptions in the code. 

It's not so much that, as that I was happy to get something halfway reliable to work at all on top of this mess of an interface.

So you'd prefer to always get the actual text dimensions, when available? What if a position doesn't have a text node on the relevant side? What if there _is_ text, but it is part of some widget (say, an inline image with a caption, or some complicated control), is iterating into the node to reach the first text node still appropriate there?

**Comment by hedgepigdaniel** [comment]
> It's not so much that, as that I was happy to get something halfway reliable to work at all on top of this mess of an interface.

Haha, yes it doesn't seem the greatest.

> So you'd prefer to always get the actual text dimensions, when available? What if a position doesn't have a text node on the relevant side? What if there _is_ text, but it is part of some widget (say, an inline image with a caption, or some complicated control), is iterating into the node to reach the first text node still appropriate there?

For this use case I think that would work. I'm not aware of all the implications, so maybe whats more useful is to explain what I'm trying to do and why it isn't working:

We have inline text in a ProseMirror document, and there are inline decorations around some of it which render something resembling a highlight (like with a highlight marker). These are rendered with `background-color`. To make the highlighted area as tall as we want it to be, the inline decorations have vertical padding.

Separately, when the cursor is placed at a point in the text, we render an element (outside of prosemirror). The location of the element is calculated using the output of `posAtCoords`.

The issue is that the output of `posAtCoords` doesn't return consistent results relative to the position of the text. Sometimes it's relative to the text, other times its relative to the inline decoration element including it's padding. It depends on whether the cursor position lies outside, within, or on the leading/trailing edge of the highlighted sections of text. https://github.com/ProseMirror/prosemirror-view/commit/250d4337216b5c24ccd4a847cc8e29c13694a33b changed the behaviour but it is still not consistent.

---

Based on what I know, my instinct would be:
 - when called for a position/side that points to a ProseMirror text node, `posAtCoords` should return the vertical position of the relevant side of the corresponding HTML text node, not of any elements surrounding it.
 - HTML nodes used to render decorations should be ignored in `posAtCoords`. e.g. if there is a widget decoration, `posAtCoords` should skip any DOM nodes used to render it and refer to the position of the DOM node used to render the ProseMirror node.

What do you think?

**Comment by marijnh** [comment]
Could you see if patches 0c5a03c...7a0a70c help with your use case?

**Comment by hedgepigdaniel** [comment]
Thanks, unfortunately those patches they don't fix the problem.

Here's a reproduction in case it helps: https://glitch.com/edit/#!/sandy-oceanic-parsley

If you move the cursor around, a red border is drawn around the result of `posAtCoords(state.selection.$head.pos)`, and its also logged to the console. The problem is that the vertical part of the position is not consistent relative to the text.

**Comment by marijnh** [comment]
The script on the project website that that Glitch was loading had (for reasons of flaky scripts) not been updated for ProseMirror-view 0.17. I've updated it manually now, and I don't see the problem anymore.

**Comment by hedgepigdaniel** [comment]
ooft, 6 days and the reproduction is already deleted 😢 

Recreated here: https://glitch.com/edit/#!/equinox-evergreen-gasoline

The issue still occurs for me on 1.17.0 but is fixed on 1.17.2. So I suspect https://github.com/ProseMirror/prosemirror-view/commit/206b48171f242016da734eaf2173423ccfb99a29 fixed it.

Thanks!
